lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -602,41 +608,17 @@ func (c *Client) handleBitmap(cb *C.CGOBitmap) C.CGOErrCode {
 	uptr := (*uint8)(ptr)
 	data := unsafe.Slice(uptr, int(cb.data_len))
 
-	// Convert BGRA to RGBA. It's likely due to Windows using uint32 values for
-	// pixels (ARGB) and encoding them as big endian. The image.RGBA type uses
-	// a byte slice with 4-byte segments representing pixels (RGBA).
-	//
-	// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/8ab64b94-59cb-43f4-97ca-79613838e0bd
-	//
-	// Also, always force Alpha value to 100% (opaque). On some Windows
-	// versions (e.g. Windows 10) it's sent as 0% after decompression for some reason.
-	for i := 0; i < len(data); i += 4 {
-		data[i], data[i+2], data[i+3] = data[i+2], data[i], 255
-	}
-
-	rect := image.Rectangle{
-		Min: image.Pt(int(cb.dest_left), int(cb.dest_top)),
-		Max: image.Pt(int(cb.dest_right)+1, int(cb.dest_bottom)+1),
-	}
-
-	var img *image.RGBA
-	isFullBitmap := cb.dest_right-cb.dest_left == 63 &&
-		cb.dest_bottom-cb.dest_top == 63
-	if isFullBitmap {
-		if c.img == nil {
-			c.img = image.NewRGBA(rect)
-		} else {
-			c.img.Rect = rect
-		}
-		img = c.img
-	} else {
-		img = image.NewRGBA(rect)
-	}
-
-	copy(img.Pix, data)
-
-	if err := c.cfg.Conn.WriteMessage(tdp.NewPNG(img, c.cfg.Encoder)); err != nil {
-		c.cfg.Log.Errorf("failed to send PNG frame %v: %v", img.Rect, err)
+	c.png2FrameBuffer = c.png2FrameBuffer[:0]
```

Comment:
clearly the new shared `png2FrameBuffer` is not thread-safe, I'm not too familiar with this part of the code, is there a strong guarantee that `handlePNG` will not be called concurrently?

Reply:
Yeah. The shared PNG encoder that we used to use from Go was also not thread safe. It's always invoked via callback on the same goroutine.

Reply:
@nklaassen essentially yes, we've built this in a way that `handlePNG` will never be called concurrently.

Brings up a potentially interesting potential point of performance improvement, though. If the RDP server is in fact sending us back png frames faster than we can dispatch them to the client, we might be able to squeeze out more performance here by making this multi-threaded. cc @LKozlowski @zmb3